ouroboros-network-framework/src/Ouroboros/Network/ConnectionManager/Core.hs

```diff
@@ -615,15 +615,24 @@ withConnectionManager ConnectionManagerArguments {
           return (freshIdSupply, v)
 
     let readState
-          :: m (Map peerAddr AbstractState)
-        readState =
-          atomically $ do
-              state <- readTMVar stateVar
-              traverse ( fmap (abstractState . Known)
-                       . readTVar
-                       . connVar
-                       )
-                       state
+          :: STM m (Map peerAddr AbstractState)
+        readState = do
+          state <- readTMVar stateVar
+          traverse ( fmap (abstractState . Known)
+                   . readTVar
+                   . connVar
+                   )
+                   state
+
+        waitForOutboundConnectionTermination
+          :: peerAddr
+          -> STM m ()
+        waitForOutboundConnectionTermination addr = do
+          state <- readState
+          case Map.lookup addr state of
+            Nothing                  -> return ()
+            Just UnknownConnectionSt -> return ()
+            Just _                   -> retry
```

Comment:
I think this is still problematic, it assumes that there's no `inbound` side using the connection.  I think this ought to be:
```hs
        waitForOutboundConnectionTermination addr = do
          state <- readState
          case Map.lookup addr state of
            Nothing                  -> return ()
            Just UnknownConnectionSt -> return ()
            Just ReservedOutboundSt  -> retry
            Just InboundIdleSt {}    -> return ()
            Just InboundSt {}        -> return ()
            Just OutboundUniSt       -> retry
            Just OutboundIdleSt      -> retry
            Just DuplexSt            -> retry
            Just WaitRemoteIdleSt    -> return ()
            Just TerminatingSt       -> retry
            Just TerminatedSt        -> return ()
```
a better name for this API would be `waitForOutboundDemotion`.  Note that it's a bit strange that we need to pattern match on both `Nothing` and `UnknownConnectionSt` (both signify the same thing). The `UnknownConnnectionSt` is probably only used for tracing, and never written to the state.

Reply:
> Note that it's a bit strange that we need to pattern match on both Nothing and UnknownConnectionSt (both signify the same thing). The UnknownConnnectionSt is probably only used for tracing, and never written to the state.

Yes I think that's true now that you mention it!

I still don't get why you think that. If the inbound side is still using the connecction then the connection state is not going to get terminated. It will transition to some inbound state.

I think it should be:

```
        waitForOutboundConnectionTermination addr = do
          state <- readState
          case Map.lookup addr state of
            Nothing                  -> return ()
            Just UnknownConnectionSt -> return ()
            Just TerminatedSt        -> return ()
			Just _                   -> retry
```

This assuming I am right and it doesn't matter if we care only about outbound states. Also remember that this is will be called in the `peerMonitoringLoop` that mostly checks for async problems like mux/miniprotocol failures

Reply:
In the outbound governor we want to wait until we can re-promote the peer, not when the connection manager can re-establish the connection.  Consider a situation in which the `outbound-governor` calls this method while the inbound side is running just fine.  It will block until the inbound side terminates, which will disallow us to use the connection until the remote side decides to terminate it, which might be never.

Reply:
Even if this is only called when there was an error, it's better to provide something robust which will make it harder for us to introduce a bug.